### DISCUSSION TITLE
Heat Exchangers (1D)

### DISCUSSION THREAD

**Discussion by ksbeattie** [IDAES/idaes-pse#240] [Q&A] [answered]
(From an email to idaes-support/users...)

Hello,

My name is Luwis Andradi and I'm a student at Minnesota State University. Also, I'm new to IDAES. 
I am working with IDAES to build a 1d heat exchanger unit model. I found your "heat_exchanger_1D.py" file and convert it into a Jupyter notebook file(Both files attached here). However, it doesn't give any outcome. 
I have changed "zero_dimensional" to "one_dimensional" in the code. Also, I found a pdf in this link- "http://idaes.org/wp-content/uploads/sites/10/2019/07/idaes_workshop_at_focapd_booklet_online_viewing_7-12-19.pdf" try to follow the "Module 1- Modeling a single unit operation" in page 9. But didn't get to work.

[1D heat exchanger.txt](https://github.com/IDAES/idaes-pse/files/6125853/1D.heat.exchanger.txt)
[heat_exchanger_1D.ipynb.txt](https://github.com/IDAES/idaes-pse/files/6125876/heat_exchanger_1D.ipynb.txt) (remove the extra `.txt` added to allow upload here).

Is it possible to looking at Jupyter notebook and give some feedback? Appreciate it.

Thanks,
Luwis Andradi

**Comment by andrewlee94** [comment, marked as answer]
Hi Luwis,

The reason why you do not see any results is because the code you have there is a set of instructions on how to construct a 1D heat exchanger model, but not the code to actually execute those instructions and build a model. In short, the unit model classes in IDAES are simply a set of instructions, and you need to create instances of these within an IDAES flowsheet in order to be able to create an actual model.

I suggest you start my looking at the IDAES introductory tutorials, which include Module 1 you mentioned above. This shows how to set up a flowsheet and construct an instance of a unit model - in that case a Flash unit. From there, you can replace the Flash unit with a Heat Exchanger instead to see how different unit operations work - note that a heat exchanger has two inlets and outlets, as opposed ot the Flash which has one inlet and two outlets.

**Reply by ravindurra** [reply]
thanks for the idea. Actually, I'm stuck right now in setting properties and figuring out how many degrees_of_freedom in the heat exchanger 1D. Can you give some idea for that? How can we exactly find out what are the degrees_of_freedom? It says it has 7 but how can we get those? 
for example heater has 4 degrees of freedom.

``` # Setup the heater model by fixing the inputs and heat duty as in documentation
m.fs.heater.inlet.enth_mol.fix(55000) # steam (~ 562.44 degrees K)
#m.fs.heater.inlet.enth_mol.fix() # C02 (~ 562.44 degrees K)
m.fs.heater.inlet.flow_mol.fix(100)
m.fs.heater.inlet.pressure.fix(101325)
m.fs.heater.heat_duty.fix(h_d[0]) ```

just like that how can I find out those 7 degrees of freedom for heat exchange 1D

``` # Import objects from pyomo package 
from pyomo.environ import ConcreteModel, SolverFactory, value
from pyomo.opt import TerminationCondition, SolverStatus

# Import the main FlowsheetBlock from IDAES. The flowsheet block will contain the unit model
from idaes.core import FlowsheetBlock

# Import idaes logger to set output levels
import idaes.logger as idaeslog

# Import the IDAES IAPWS-95 Steam properties package to create a properties block for the flowsheet
from idaes.generic_models.properties import iapws95
# Import the IDAES Span-Wager CO2 properties package to create a properties block for the flowsheet
from idaes.generic_models.properties import swco2

# Import heater unit model from the model library
#from idaes.generic_models.unit_models.heater import Heater
#from idaes.generic_models.unit_models.heat_exchanger_1D import HeatExchanger1D

#from idaes.generic_models.unit_models.heat_exchanger_1D import HeatExchanger1DData

# Import the degrees_of_freedom function from the idaes.core.util.model_statistics package
# DOF = Number of Model Variables - Number of Model Constraints
from idaes.core.util.model_statistics import degrees_of_freedom

# Import matplotlib for create charts
import matplotlib.pyplot as plt



# Create the ConcreteModel and the FlowsheetBlock, and attach the flowsheet block to it.
m = ConcreteModel()
# dynamic or ss flowsheet needs to be specified here
m.fs = FlowsheetBlock(default={"dynamic": False}) 

#At this point, we have a single Pyomo model that contains an (almost) empty flowsheet block.
# Todo: call pprint on the model
m.pprint()


# Add IDAES IAPWS-95 Steam properties package "Iapws95ParameterBlock" parameter block to the flowsheet with specifications
m.fs.properties = iapws95.Iapws95ParameterBlock()

# Add IDAES Span-Wager CO2 properties package "SWCO2ParameterBlock" parameter block to the flowsheet with specifications
#m.fs.properties = swco2.SWCO2ParameterBlock()




# Create an instance of the heater unit, attaching it to the flowsheet
# Specify that the property package to be used with the HeatExchanger1D is the one we created earlier.

m.fs.heatexchanger1D = HeatExchanger1D(default={
        "shell_side":{"property_package": m.fs.properties},
        "tube_side":{"property_package": m.fs.properties}})

#default = 'use_parent_value'


# Call the degrees_of_freedom function, get intitial DOF
DOF_initial = degrees_of_freedom(m)
print("The initial DOF is {0}".format(DOF_initial))


help(degrees_of_freedom)```

Thanks in advance. Any help appriciate it

**Reply by andrewlee94** [reply]
Determining the correct degrees of freedom can be a difficult question to answer. For some examples of the 1D heat exchanger in use, havea look at https://github.com/IDAES/idaes-pse/blob/main/idaes/generic_models/unit_models/tests/test_heat_exchanger_1D.py. Notably, on line 677 there is an example using the IAWPS95 property package.

First and foremost, most of the degrees of freedom will be associated with the inlets to the unit model, but exactly what those are depends on the property package being used. In this case, you are using the Helmholtz equations of state (IAPWS95 and SWCO2), so you have `flow_mol`, `enth_mol` and `pressure` in the two inlets to be defined (thus 6 degrees of freedom).

The final set of degrees of freedom are the unit model design specifications - things like heat exchanger area and heat transfer coefficient. I am actually surprised to see only 1 more degree of freedom in the 1D heat exchanger model. I am not an expert on that model, but I would have expected you to need to provide values for the following:

* shell_length
* tube_length
* d_shell (diameter of shell)
* d_tube_inner and d_tube_outer (inner and outer diameter of heat exchanger tubes)
* N_tubes (number of heat exchanger tubes)
* shell_heat_transfer_coefficient
* tube_heat_transfer_coefficient

**Reply by jghouse88** [reply]
@ravindurra Please look at the docs for more details on DOF: https://idaes-pse.readthedocs.io/en/stable/technical_specs/model_libraries/generic/unit_models/heat_exchanger_1D.html?highlight=heat%20exchanger%201D. 

Also, as @andrewlee94 points out looking at the test files will be helpful to see how to use the hx1d model.

**Reply by ravindurra** [reply]
Thanks all
